### ISSUE TITLE
Wallet backend running with file db crashes on integration tests

### ISSUE THREAD

**Issue by piotr-iohk** [cardano-foundation/cardano-wallet#1271]
# Context
 
 

| Information  | -                                                     |
| ---          | ---                                                   |
| Version      | `v2020.1.7` |
| Platform     | MacOS, Linux (might be others)          |
| Installation | Daedalus, standalone           |

Problem reported by @darko-mijic. Observed when running Daedalus automated integration tests.
(Note that Daedalus integration tests use wallet backend with db that is file based)
Similar also observed when running wallet backend integration tests with file database instead of in-memory database -> https://github.com/input-output-hk/cardano-wallet/pull/1240

# Steps to Reproduce
Note that wallet does not crash on any specific scenario. The crash is observed randomly while tests are executed only if wallet is using database on filesystem. Our integration tests running with in-memory db work fine.

**Using Daedalus:**

```
git clone https://github.com/input-output-hk/daedalus.git
cd daedalus
git checkout -b v2-integration origin/v2-integration
yarn install
yarn build
yarn nix:dev
yarn test:e2e
```
:point_up_2:  executes Daedalus integration tests which produce a lot of requests to wallet backend in a short time. After a while of tests being executed cardano-wallet _crashes_. (cardano-wallet process is no longer listed on `ps -ef`)
In the log (`~/.local/share/Daedalus/selfnode/Logs/pub/node.log`) there is:

```
StatementAlreadyFinalize[cardano-wallet.worker:Info:2779] [2020-01-16 09:55:54.34 UTC] 65c2adea: Applying blocks [76554.148 ... 76554.148]
d "BEGIN"
[cardano-wallet.worker:Error:2779] [2020-01-16 09:55:54.34 UTC] 65c2adea: Recoverable error following the chain: StatementAlreadyFinalized "BEGIN"
Jan 16 10:55:55.001 INFO Leader event started, event_remaining_time: 1s 998ms 803us 580ns, event_end: 2020-01-16T09:55:57+00:00, event_start: 2020-01-16T09:55:55+00:00, event_date: 76554.149, leader_id: 1, task: leadership
Jan 16 10:55:55.001 INFO receiving block from leadership service, date: 76554.149, parent: f87cb7aa6d08dbb8535526d09711c6964984d6da0d57df0f61131b2cb78ab7de, hash: 2ee568c40c80c135df5d5c37c3f55adf9ec0e27a76d508f6a46a87442c9fb138, task: block
Jan 16 10:55:55.017 INFO block from leader event successfully stored, date: 76554.149, parent: f87cb7aa6d08dbb8535526d09711c6964984d6da0d57df0f61131b2cb78ab7de, hash: 2ee568c40c80c135df5d5c37c3f55adf9ec0e27a76d508f6a46a87442c9fb138, task: block
Jan 16 10:55:55.017 INFO update current branch tip, date: 76554.149, parent: f87cb7aa6d08dbb8535526d09711c6964984d6da0d57df0f61131b2cb78ab7de, hash: 2ee568c40c80c135df5d5c37c3f55adf9ec0e27a76d508f6a46a87442c9fb138, task: block
[cardano-wallet.stake-pools:Info:27] [2020-01-16 09:55:55.39 UTC] Applying blocks [76554.149 ... 76554.149]
[cardano-wallet.stake-pools:Info:27] [2020-01-16 09:55:55.39 UTC] Writing stake-distribution for epoch 76554
malloc(): memory corruption
```
Then `cardano-wallet` tries to start again but:
```
[cardano-wallet:Info:4] [2020-01-16 09:55:56.63 UTC] Using directory: /home/piotr/.local/share/Daedalus/selfnode
[cardano-wallet:Info:4] [2020-01-16 09:55:56.63 UTC] Using directory: /home/piotr/.local/share/Daedalus/selfnode/wallets
[cardano-wallet:Info:4] [2020-01-16 09:55:56.63 UTC] Running as v2020.1.7 (git revision: 0000000000000000000000000000000000000000)
[cardano-wallet:Info:4] [2020-01-16 09:55:56.63 UTC] Wallet backend server starting...
[cardano-wallet:Info:4] [2020-01-16 09:55:56.63 UTC] Node is Jörmungandr on testnet
[cardano-wallet:Alert:4] [2020-01-16 09:55:56.63 UTC] The API server listen port 8088 is already in use.
```

**Using only wallet backend:**
 - Run integration tests with changes from #1240. 
 - the error is quite similar:
```
[cardano-wallet.worker:Error:4762] [2019-12-31 12:34:49.81 UTC] 7f084053: Recoverable error following the chain: SQLite3 returned ErrorMisuse while attempting to perform bind text.
Statement[cardano-wallet.worker:Info:4762] [2019-12-31 12:34:50.36 UTC] 7f084053: Applying blocks [1079666.5 ... 1079677.5]
AlreadyFinalized "BEGIN"
[cardano-wallet.worker:Error:4762] [2019-12-31 12:34:50.36 UTC] 7f084053: Recoverable error following the chain: StatementAlreadyFinalized "BEGIN"
[cardano-wallet.stake-pools:Info:2025] [2019-12-31 12:34:51.17 UTC] Applying blocks [1079681.7 ... 1079681.7]
malloc(): memory corruption
```

## Expected behavior

Wallet backend does not crash.

## Actual behavior
 
It crashes.

---

# Resolution 

---

# QA

**Comment by KtorZ** [comment]
Most likely fixed as part of #1268 , our integration tests now also run using file database.

**Comment by piotr-iohk** [comment]
Closing. Integration tests passing now using filesystem database on recent master.
